File: Graphs/Hard/Dijkstras Algorithm/Dijkstras Algorithm.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <unordered_map>

using namespace std;

typedef pair<int, int> Edge;  // pair representing (weight, destination vertex)

void dijkstra(const unordered_map<int, vector<Edge>>& graph, int start, unordered_map<int, int>& distances, unordered_map<int, int>& predecessors) {
    priority_queue<Edge, vector<Edge>, greater<Edge>> pq;  // Min-heap to get the minimum distance

    distances[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        int current_vertex = pq.top().second;
        int current_distance = pq.top().first;
        pq.pop();

        // If current distance is greater than the known distance, skip
        if (current_distance > distances[current_vertex]) {
            continue;
        }

        // Explore neighbors
        for (const Edge& neighbor : graph.at(current_vertex)) {
            int neighbor_vertex = neighbor.second;
            int weight = neighbor.first;
            int distance = current_distance + weight;

            // If a shorter path is found, update the distance and predecessor
            if (distance < distances[neighbor_vertex]) {
                distances[neighbor_vertex] = distance;
                predecessors[neighbor_vertex] = current_vertex;
                pq.push({distance, neighbor_vertex});
            }
        }
    }
}

void printPath(const unordered_map<int, int>& predecessors, int target) {
    vector<int> path;
    int current = target;
    while (current != -1) {
        path.insert(path.begin(), current);
        current = predecessors.at(current);
    }

    cout << "Shortest path: ";
    for (int vertex : path) {
        cout << vertex << " -> ";
    }
    cout << "\b\b  \n";
}

int main() {
    unordered_map<int, vector<Edge>> graph;

    // Taking user-defined input for the graph
    int numEdges;
    cout << "Enter the number of edges: ";
    cin >> numEdges;

    cout << "Enter the edges (source destination weight):\n";
    for (int i = 0; i < numEdges; ++i) {
        int source, destination, weight;
        cin >> source >> destination >> weight;
        graph[source].push_back({weight, destination});
        // Assuming an undirected graph
        graph[destination].push_back({weight, source});
    }

    int startNode, endNode;
    cout << "Enter the start node: ";
    cin >> startNode;
    cout << "Enter the end node: ";
    cin >> endNode;

    // Initialize distances and predecessors
    unordered_map<int, int> distances, predecessors;
    for (const auto& pair : graph) {
        distances[pair.first] = INT_MAX;
        predecessors[pair.first] = -1;  // -1 represents no predecessor
    }

    dijkstra(graph, startNode, distances, predecessors);

    if (distances[endNode] != INT_MAX) {
        cout << "Shortest distance from " << startNode << " to " << endNode << ": " << distances[endNode] << endl;
        printPath(predecessors, endNode);
    } else {
        cout << "There is no path from " << startNode << " to " << endNode << "." << endl;
    }

    return 0;
}

File: Graphs/Hard/Dijkstras Algorithm/Dijkstras Algorithm.java
import java.util.*;

public class DijkstraAlgorithm {

    static class Edge {
        int destination;
        int weight;

        public Edge(int destination, int weight) {
            this.destination = destination;
            this.weight = weight;
        }
    }

    public static void dijkstra(Map<Integer, List<Edge>> graph, int start, Map<Integer, Integer> distances, Map<Integer, Integer> predecessors) {
        PriorityQueue<Edge> priorityQueue = new PriorityQueue<>(Comparator.comparingInt(e -> e.weight));
        priorityQueue.offer(new Edge(start, 0));

        distances.put(start, 0);

        while (!priorityQueue.isEmpty()) {
            Edge currentEdge = priorityQueue.poll();
            int currentVertex = currentEdge.destination;

            if (currentEdge.weight > distances.get(currentVertex)) {
                continue;
            }

            for (Edge neighbor : graph.getOrDefault(currentVertex, new ArrayList<>())) {
                int newDistance = distances.get(currentVertex) + neighbor.weight;

                if (newDistance < distances.get(neighbor.destination)) {
                    distances.put(neighbor.destination, newDistance);
                    predecessors.put(neighbor.destination, currentVertex);
                    priorityQueue.offer(new Edge(neighbor.destination, newDistance));
                }
            }
        }
    }

    public static void printPath(Map<Integer, Integer> predecessors, int target) {
        List<Integer> path = new ArrayList<>();
        int current = target;

        while (current != -1) {
            path.add(current);
            current = predecessors.get(current);
        }

        Collections.reverse(path);
        System.out.print("Shortest path: ");
        for (int vertex : path) {
            System.out.print(vertex + " -> ");
        }
        System.out.println("\b\b  ");
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        Map<Integer, List<Edge>> graph = new HashMap<>();

        System.out.print("Enter the number of edges: ");
        int numEdges = scanner.nextInt();

        System.out.println("Enter the edges (source destination weight): ");
        for (int i = 0; i < numEdges; ++i) {
            int source = scanner.nextInt();
            int destination = scanner.nextInt();
            int weight = scanner.nextInt();

            graph.computeIfAbsent(source, k -> new ArrayList<>()).add(new Edge(destination, weight));
            // Assuming an undirected graph
            graph.computeIfAbsent(destination, k -> new ArrayList<>()).add(new Edge(source, weight));
        }

        System.out.print("Enter the start node: ");
        int startNode = scanner.nextInt();
        System.out.print("Enter the end node: ");
        int endNode = scanner.nextInt();

        Map<Integer, Integer> distances = new HashMap<>();
        Map<Integer, Integer> predecessors = new HashMap<>();

        // Initialize distances and predecessors
        for (int vertex : graph.keySet()) {
            distances.put(vertex, Integer.MAX_VALUE);
            predecessors.put(vertex, -1); // -1 represents no predecessor
        }

        dijkstra(graph, startNode, distances, predecessors);

        if (distances.get(endNode) != Integer.MAX_VALUE) {
            System.out.println("Shortest distance from " + startNode + " to " + endNode + ": " + distances.get(endNode));
            printPath(predecessors, endNode);
        } else {
            System.out.println("There is no path from " + startNode + " to " + endNode + ".");
        }
    }
}
